test/core/test_lwt_list.ml

```diff
@@ -119,6 +119,96 @@ let test_map f test_list =
   in
   ()
 
+let test_map_parallelism ?(rev=false) map =
+  let t, w = Lwt.wait () in
+  let g x =
+    Lwt.wakeup_later w ();
+    Lwt.return (x + 1) in
+  let f x =
+    if x = 0 then t >>= (fun _ -> Lwt.return (x + 1))
+    else g x
+  in
+  let p = map f [0; 1] in
+  let expected = if rev then [2; 1] else [1; 2] in
+  p >>= (fun l -> Lwt.return (l = expected))
+
+let test_map_serialism ?(rev=false) map =
+  let m = Lwt_mutex.create () in
+  let g x =
+    Lwt_mutex.unlock m;
+    Lwt.return (x + 1) in
+  let f x =
+    Lwt_mutex.lock m >>= fun _ ->
+    if x = 0 then g x
+    else Lwt.return (x + 1)
+  in
+  let p = map f [0; 1] in
+  let expected = if rev then [2; 1] else [1; 2] in
+  p >>= (fun u -> Lwt.return (u = expected))
```

Comment:
This test doesn't seem to be testing serialization of calls to `f` to me. `map_s` is supposed to guarantee that `f 1` isn't called until the promise returned by `f 0` is fulfilled.

For this test to check that, it would have to have some way of failing if `f 1` starts running before the promise `f 0` is fulfilled.

I'm actually not sure immediately how to do that. Perhaps you would have to mutate some state in a callback of `f 0`, and `f 1` would be fulfilled with that state. I don't think that's legitimate, though, because it would depend on Lwt's internal ordering of callback calls to ensure that the `f 0` callback that mutates the state runs before the call to `f 1` scheduled by `map_s`.

Reply:
`Lwt.pause () >>= f` might help here. It will force everything that is ready to run immediately to run before `f`.

Reply:
I've posted this new approach:
```
let test_map_serialism map =
  let l = ref [] in
  let f x =
    let p = Lwt.pause () >>= fun _ -> Lwt.return x in
    assert (List.length !l = 0 || not @@ Lwt.is_sleeping @@ List.hd !l);
    l := p :: !l;
    p
  in
  let p' = map f [0; 1; 2] in
  p' >>= (fun _ -> Lwt.return_true)
```
This test passes if I give it `Lwt_list.map_s`, but fails if I give it `Lwt_list.map_p`. I'm not sure if this totally captures what you were asking for above, though.

Reply:
I was thinking something much simpler. I just tried this:

```ocaml
let test_map_serialism map =
  let f2_ran = ref false in

  let f x =
    if x = 1 then
      Lwt.pause () >>= fun () ->
      assert (not !f2_ran);
      Lwt.return ()
    else begin
      f2_ran := true;
      Lwt.return ()
    end
  in

  map f [1; 2] >>= fun _result ->
  Lwt.return true
```

The state is just a boolean. `f 1` starts and puts itself to sleep until the next Lwt "tick." This should give `f 2` a chance to run if it can, and if it does, it will flip the state. So when `f 1` is resumed on the next tick, we just check that the state still hasn't been touched. At the end, for good measure, we could check the result is correct, but that overlaps with other tests already written.

This also passes with `map_s`, but not `map_p`. I didn't add parameters for working with `_rev` variants.